File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public int GoodNodes(TreeNode root) {
       return GoodNodes(root, root.val);
    }

    private int GoodNodes(TreeNode root, int max) {
        if (root == null) return 0;

        int left = GoodNodes(root.left, Math.Max(max, root.val));
        int right = GoodNodes(root.right, Math.Max(max, root.val));

        if (max > root.val) return left + right;
        return left + right + 1;
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int result = 0;
        dfs(root, result);
        return result;
    }

    int dfs(TreeNode* root, int& diameter) {
        if (root == nullptr) return 0;

        int left = dfs(root->left, diameter);
        int right = dfs(root->right, diameter);

        int path = left + right;
        if (path > diameter) diameter = path;

        return std::max(left, right) + 1;
    }
};
